### DISCUSSION TITLE
The price of tightness and compactness in SpineOpt

### DISCUSSION THREAD

**Discussion by manuelma** [spine-tools/SpineOpt.jl#972] [Ideas]
Running some benchmarks I realized the unit flow capacity constraint is taking a lot of the build time. I tried removing the 'complex' terms that make it tight and compact and it's of course much faster to build, but solving time isn't that much longer in the benchmarks I run. The result also didn't change. In other words, by making the constraint tight and compact we are adding to the execution time instead of subtracting from it. This is probably because SpineOpt has a harder time in computing certain indices compared to the flat deterministic model that the tight formulation was originally designed for.

So I suggest simplifying the constraint untill we figure out a better way to implement tight and compact constraints in SpineOpt. Doing it will reduce the execution time by about 20% in the benchmarks I'm running.

Maybe this should be a discussion @clizbe ?
Also, I might be using the tight and compact words incorrectly but hopefully you'll get what I mean.

**Comment by datejada** [comment]
The original idea was to include a parameter/method that triggers different versions of the constraints. This is what @g-moralesespana proposed with the different levels of detail for SpineOpt.
Last year's changes aimed to have the model's most complex version (Tight and Compact formulation). However, we didn't tackle the different levels of detail at that time.
Honestly, I didn't expect such an impact in creating the constraint since we have experience working in other languages (e.g., GAMS and Python). But, since the evidence is there, SpineOpt is slow to build this constraint. Let's go forward now with a parameter that triggers the constraint you want. By default, we use a simple version, but users can select the Tight and Compact version if they want to.
The benefits of the Tight and Compact formulation will highly depend on the input data (unit characteristics, demand, renewable production, and so on). In cases where the unit commitment constraints are not binding (like the one you are mentioning -> no change in the objective function), you don't see the benefit of more complex formulations in your model. However, if the case study heavily uses unit commitment constraints, they are key to performance.
We can discuss it in the next developers' meeting to ensure we agree on this approach.

**Comment by manuelma** [comment]
Thank you @datejada that makes sense. I just wanted to add that it might not be a matter of the language but the specific application, SpineOpt, where the next time step is not unique, and the minimum up time can be stochastic, and so on, which makes it more difficult to write the constraint. I don't know if you tried to build the tight and compact constraint in GAMS or Python with similar assumptions as in SpineOpt, if you did then we could probably try and learn from that implementation?

**Reply by datejada** [reply]
We have experience with these constraints without all the temporal complexity that SpineOpt has. So, it is not a fair comparison. Depending on the general improvements to speed the building of the constraints in SpineOpt, we can also benefit from those changes to build the tight and compact ones (when the user chooses it), as @tarskul comments below.

**Comment by tarskul** [comment]
The flexible time steps are indeed different. Adding a few terms can therefore impact the build time. However, as we have discussed before, it should not make that much of a difference. I would even argue that we should be able to implement the tight and compact formulation without the need for the toggle to the simple formulation. Regardless, at some point we need to tackle the building of the constraints itself. I have some ideas that I want to test but that is going to take a while before I can get to that.

**Reply by manuelma** [reply]
Maybe you could share some of those ideas @tarskul ? I'm curious.

Regarding the tight and compact formulations I agree we should be able to implement them and we have been able to do it, but it looks like not all models will benefit from them, and a lot of models will actually perform better without them. So it would make sense to leave the choice to the user on which formulation to use. I don't think we can determine analytically which one would be better for each model.

**Reply by tarskul** [reply]
First I'd like to do at least some very basic (and theoretical) initial internal testing such that I know myself what I'm talking about (the cookies are still in the oven). I don't want to waste everybody's time with nonsense (if I burn the cookies). I'll try to prioritise the minimum I need for those tests such that it doesn't take months (for you to taste the first batch of cookies).
